lib/iris/fileformats/_nc_load_rules/actions.py

```diff
@@ -0,0 +1,337 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+"""
+Replacement code for the Pyke rules.
+
+For now, we are still emulating various aspects of how our original Pyke-based
+code used the Pyke 'engine' to hold translation data, both Pyke-specific and
+not :
+1) basic details from the iris.fileformats.cf analysis of the file are
+   recorded before translating each output cube, using
+   "engine.assert_case_specific_fact(name, args)".
+
+2) this is also used to store intermediate info passed between rules, which
+   used to be done with a "facts_cf.provides" statement in rule actions.
+
+3) Iris-specific info is (still) stored in additional properties created on
+   the engine object :
+       engine.cf_var, .cube, .cube_parts, .requires, .rule_triggered, .filename
+
+Our "rules" are just action routines.
+The top-level 'run_actions' routine decides which actions to call, based on the
+info recorded when processing each cube output.  It does this in a simple
+explicit way, which doesn't use any clever chaining, "trigger conditions" or
+other rule-type logic.
+
+Each 'action' function can replace several similar 'rules'.
+E.G. 'action_provides_grid_mapping' replaces all 'fc_provides_grid+mapping_<X>'.
+To aid debug, each returns a 'rule_name' string, indicating which original rule
+this particular action call is emulating :  In some cases, this may include a
+textual note that this rule 'failed', aka "did not trigger", which would not be
+recorded in the original implementation.
+
+The top-level 'run_actions' ensures that the individual rules actions are
+called, with various arguments, as appropriate to ensure the whole cube is
+built as it was by the original rules implementation.
+
+TODO: remove the use of intermediate "facts" to carry information between
+actions.  This mimics older behaviour, so is still useful while we are still
+comparing behaviour with the old Pyke rules (debugging).  But once that is no
+longer useful, this can be considerably simplified.
+
+"""
+
+from . import helpers as hh
+from functools import wraps
+
+
+def convert_actionname_to_rulename(func_name):
+    # Given the name of an action-func, return the name of the rule.
+    funcname_prefix = "action_"
+    rulename_prefix = "fc_"  # To match existing behaviours
+    rule_name = func_name
+    if rule_name.startswith(funcname_prefix):
+        rule_name = rule_name[len(funcname_prefix) :]
+    if not rule_name.startswith(rulename_prefix):
+        rule_name = rulename_prefix + rule_name
+    return rule_name
+
+
+def _default_rulenamesfunc(func_name):
```

Comment:
This seems to be functionally a copy of `convert_action_to_rulename`. Do there need to be two different functions here?

Reply:
You're right :+1: 
This has been through several phases of rework + got simplified, and probably isn't done yet (still developing).
I **_might_** still decide that all actions should simply return a rule-name-like-string, and then no wrapper is needed.
At present, that is looking simpler.

I'd like to leave this conversation unresolved + come back to it when I've done more dev on the actions.
(my  present focus is on getting tests for everything)

Reply:
Ok, I have now simplified, but have not removed the wrapper / default-name mechanisms
(in https://github.com/SciTools/iris/pull/4170/commits/b26fbcd79e3e34d10f122cbadab46123348830c0)